locksmith/src/operations/receiptOperations.ts

```diff
@@ -1,52 +1,119 @@
 import { SubgraphService } from '@unlock-protocol/unlock-js'
 import { Receipt, ReceiptBase } from '../models'
+import { getMetadata } from './userMetadataOperations'
+import { lowercaseObjectKeys } from '../utils/object'
 
 interface ReceiptDetailsProps {
   lockAddress: string
   network: number
   hash: string
+  tokenId?: string
 }
 
-export const getReceiptDetails = async ({
+export const getSingleReceiptDetails = async ({
+  hash,
+  network,
+}: {
+  hash: string
+  network: number
+}) => {
+  // get receipts details from subgraph
+  const subgraph = new SubgraphService()
+  const receipt = await subgraph.receipt(
+    {
+      where: {
+        id: hash,
+      },
+    },
+    {
+      network,
+    }
+  )
+
+  return receipt
+}
+
+/** Get purchaser details from metadata when receipt purchaser is not present */
+const getPurchaserDetails = async ({
   lockAddress,
   network,
   hash,
-}: ReceiptDetailsProps): Promise<{
-  supplier: ReceiptBase | null
-  purchaser: Receipt | null
-  receipt: any
-}> => {
-  const purchaser = await Receipt.findOne({
+}: ReceiptDetailsProps): Promise<Partial<
+  Receipt & { email?: string }
+> | null> => {
+  const purchaserDetails = await Receipt.findOne({
     where: {
       network,
       lockAddress,
       hash,
     },
   })
 
-  const supplier = await ReceiptBase.findOne({
+  // get purchaser metadata details
+  let metadata: any
+  const receipt = await getSingleReceiptDetails({ hash, network })
+  if (receipt?.payer) {
+    metadata = await getMetadata(lockAddress, receipt.payer, true)
```

Comment:
Can you move that into the block `if (!purchaserDetails) {` ?

Reply:
moved inside the block